timemachine/fe/rbfe.py

```diff
@@ -0,0 +1,378 @@
+        ligand_conf = st.combine_confs(mol_a_conf, mol_b_conf)
+        combined_conf = np.concatenate([host_conf, ligand_conf])
+        x0 = combined_conf
+        v0 = np.zeros_like(x0)
+        box0 = host_config.box
+        group_idxs = get_group_indices(get_bond_list(hgs.bond))
+        seed = np.random.randint(0, 1000)
+        combined_masses = np.concatenate([host_masses, st.combine_masses()])
+        frames, boxes = sample(bound_impls, combined_masses, x0, v0, box0, lamb, group_idxs, seed, n_frames=n_frames)
+        all_frames.append(frames)
+        all_boxes.append(boxes)
+
+    return all_frames, all_boxes, all_U_fns
+
+
+def plot_BAR(dG, dG_err, fwd_delta_u, rev_delta_u, title, plot_idx):
+    """
+    Generate a subplot showing overlap for a particular pair of delta_us.
+    """
+    plt.subplot(*plot_idx)
+    plt.title(f"{title}, dG: {dG:.2f} +- {dG_err:.2f} kTs")
+    plt.hist(fwd_delta_u, alpha=0.5, label="fwd", density=True, bins=20)
+    plt.hist(-rev_delta_u, alpha=0.5, label="-rev", density=True, bins=20)
+    plt.xlabel("work (kTs)")
+    plt.legend()
+
+
+SimulationResult = namedtuple("SimulationResult", ["all_dGs", "all_errs", "plot_png", "frames", "boxes", "U_fns"])
+
+
+def estimate_free_energy_given_samples(all_frames, all_boxes, all_U_fns, lambda_schedule, prefix):
+    """
+    Estimate free energies given pre-generated samples.Let L be the number of lambda windows,
+    T be the number of samples, N be the number of atoms, P be the number of components
+    in the energy function.
+
+    Parameters
+    ----------
+    all_frames: np.array of shape (L, T, N, 3)
+        List of coordinates from all windows
+
+    all_boxes: np.array of shape (L, T, 3, 3)
+        List of boxes from all windows
+
+    all_U_fns: list of energy functions of shape (L, P)
+        List of energy functions from all windows
+
+    lambda_schedule: np.array of shape (L,)
+        Lambda schedule
+
+    prefix: str
+        A prefix to append to overlap figures
+
+    Return
+    ------
+    3-tuple of dGs, dG_errs, overlap_plot
+        All components have have shape (L-1,)
+
+    """
+    # assume pair-BAR format
+    # setup U_fns to operate in batch mode
+    kT = BOLTZ * 300.0
+    beta = 1 / kT
+
+    all_dGs = []
+    all_errs = []
+
+    all_batch_U_fns = []
+    for lamb_idx, lamb in enumerate(lambda_schedule):
+        bound_impls = [p.bound_impl(np.float32) for p in all_U_fns[lamb_idx]]
+        batch_U_fns = get_batch_U_fns(bound_impls, lamb)
+        all_batch_U_fns.append(batch_U_fns)
+
+    U_names = []
+    for U_fn in all_U_fns[0]:
+        # convert from '<timemachine.lib.potentials.Nonbonded object at 0x7f7880b900b8>' -> Nonbonded
+        U_names.append(repr(U_fn).split(".")[-1].split()[0])
+
+    num_rows = len(lambda_schedule) - 1
+    num_cols = len(U_names) + 1
+
+    plot_iota = 1
+
+    def get_plot_idx():
+        nonlocal plot_iota  # what is this black magic? needed to deal with += increment
```

Comment:
😮

Reply:
I think this could be avoided with something like
```python
_, axss = plt.subplots(num_rows, num_cols)
# …
for lamb, axs in zip(lambda_schedule, axss):
    # …
    for prev_U_fn, cur_U_fn, ax in zip(prev_U_fns, cur_U_fns, axs):
        # …
        plot_BAR(df, df_err, fwd_delta_u, rev_delta_u, U_names[u_idx], ax=ax)
```
(and update `plot_BAR` to take an `ax` argument instead of `plot_idx`)

Reply:
does that actually line-up? since `num_rows = len(lambda_schedule) - 1` I think this would result in an empty row?

Reply:
(also the main problem is due to a limitation in python not allowing us to modify closures iirc, and nonlocal is a patch, similar to global)

Reply:
> does that actually line-up?

Sorry, the example might need a little tweaking (maybe `prev_lambda, cur_lambda in zip(lambda_schedule, lambda)`), but my main point was that I think a nonlocal variable to keep track of which axis can be avoided by explicitly looping over the axes (e.g. by zipping with the other loop variables)

Reply:
this is a little annoying, a matplotlib `Axes` isn't the same as that of `title` in weird ways, for example:

`plt.title("my title")` is fine
`ax.title("my title")` breaks, and you'd have to use `ax.set_title("my title")` instead

Reply:
`plt.set_title` is also not a valid option either!

Reply:
Yeah, it is a bit annoying. My rule of thumb is to prepend `set_` when working with axis objects 🤷‍♂️

Reply:
>this is a little annoying, a matplotlib axes isn't the same as that of plt in weird ways, for example:
>ax.title("my title") breaks, and you'd have to use ax.set_title("my title") instead
>plt.set_title is also not a valid option either!
>(similarly for xlabel vs set_xlabel)

This section in the matplotlib documentation may be helpful re: differences between "pyplot style" and "OO style" https://matplotlib.org/stable/tutorials/introductory/usage.html#the-object-oriented-and-the-pyplot-interfaces

(and following section I think describes the current use case https://matplotlib.org/stable/tutorials/introductory/usage.html#making-a-helper-functions )

Reply:
sorted this out - uses axes now